molgenis-searchall/src/main/java/org/molgenis/searchall/controller/SearchAllController.java

```diff
@@ -0,0 +1,118 @@
+import static org.molgenis.data.system.model.RootSystemPackage.PACKAGE_SYSTEM;
+import static org.molgenis.searchall.controller.SearchAllController.BASE_URI;
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+import static org.springframework.web.bind.annotation.RequestMethod.GET;
+
+@Controller
+@RequestMapping(BASE_URI)
+public class SearchAllController
+{
+	public static final String BASE_URI = "/api/searchall/";
+	private final DataService dataService;
+
+	@Autowired
+	public SearchAllController(DataService dataService)
+	{
+		this.dataService = requireNonNull(dataService);
+
+	}
+
+	@RequestMapping(value = "/search", method = GET, produces = APPLICATION_JSON_VALUE)
+	@ResponseBody
+	public Result findAll(@RequestParam(value = "term") String searchterm)
+	{
+		List<PackageResult> packages = dataService.findAll(PackageMetadata.PACKAGE, Package.class)
+												  .filter(aPackage -> !isSystemPackage(aPackage))
+												  .filter(aPackage -> isMatchingMetadata(aPackage, searchterm))
+												  .map(pack -> PackageResult.create(pack.getLabel(),
+														  pack.getDescription()))
+												  .collect(toList());
+		List<EntityTypeResult> entityTypeMatches = new ArrayList<>();
+
+		Stream<EntityType> entityTypeStream = dataService.findAll(EntityTypeMetadata.ENTITY_TYPE_META_DATA,
+				EntityType.class).filter(entityType -> !isSystemEntity(entityType));
+
+		entityTypeStream.forEach(entityType -> searchSingleEntityType(searchterm, entityTypeMatches, entityType));
+
+		return Result.create(entityTypeMatches, packages);
+	}
+
+	private void searchSingleEntityType(String searchterm, List<EntityTypeResult> entityTypeMatches,
+			EntityType entityType)
+	{
+		boolean isLabelMatch = entityType.getLabelValue().toString().contains(searchterm);
+		boolean isDescMatch = entityType.getString(DESCRIPTION) != null ? entityType.getString(DESCRIPTION)
+																					.contains(searchterm) : false;
+		List<AttributeResult> matchingAttributes = StreamSupport.stream(entityType.getAllAttributes().spliterator(),
+				false)
+																.filter(attribute -> isMatchingMetadata(attribute,
+																		searchterm))
+																.map(attr -> AttributeResult.create(attr.getLabel(),
+																		attr.getDescription(),
+																		attr.getDataType().toString()))
+																.collect(toList());
+		long nrOfMatchingEntities = dataService.count(entityType.getId().toString(),
+				new QueryImpl<>().search(searchterm));
+		if (isLabelMatch || isDescMatch || matchingAttributes.size() > 0 || nrOfMatchingEntities > 0)
+		{
+			entityTypeMatches.add(
+					EntityTypeResult.create(entityType.getId(), entityType.getLabel(), entityType.getDescription(),
+							entityType.getPackage().getId(), isLabelMatch, isDescMatch, matchingAttributes,
+							nrOfMatchingEntities));
+		}
+	}
+
+	private boolean isMatchingMetadata(Entity metadataEntity, String term)
```

Comment:
Probably, but I'm not sure it would be an improvement. Since we're already looping over all entities, that would mean extra queries for data we already have.